tls/s2n_x509_validator.c

```diff
@@ -138,6 +138,100 @@ void s2n_x509_trust_store_wipe(struct s2n_x509_trust_store *store) {
     }
 }
 
+struct s2n_x509_cert* s2n_x509_cert_new(void) {
+    DEFER_CLEANUP(struct s2n_blob mem = { 0 }, s2n_free);
+    PTR_GUARD_POSIX(s2n_alloc(&mem, sizeof(struct s2n_x509_cert)));
+
+    struct s2n_x509_cert *cert = (struct s2n_x509_cert*)(void*) mem.data;
+    cert->cert = NULL;
+
+    ZERO_TO_DISABLE_DEFER_CLEANUP(mem);
+    return cert;
+}
+
+int s2n_x509_cert_free(struct s2n_x509_cert *cert) {
+    if (cert) {
+        POSIX_GUARD(s2n_free_object((uint8_t **) &cert, sizeof(struct s2n_x509_cert)));
+    }
+    return S2N_SUCCESS;
+}
+
+int s2n_x509_cert_get_issuer_hash(struct s2n_x509_cert *cert, unsigned long *hash) {
+    POSIX_ENSURE_REF(cert);
+    POSIX_ENSURE_REF(cert->cert);
+
+    unsigned long temp_hash = X509_issuer_name_hash(cert->cert);
+    POSIX_ENSURE(temp_hash != 0, S2N_ERR_INTERNAL_LIBCRYPTO_ERROR);
+
+    *hash = temp_hash;
+
+    return S2N_SUCCESS;
+}
+
+static struct s2n_x509_crl* s2n_x509_crl_new(void) {
+    DEFER_CLEANUP(struct s2n_blob mem = { 0 }, s2n_free);
+    PTR_GUARD_POSIX(s2n_alloc(&mem, sizeof(struct s2n_x509_crl)));
+
+    struct s2n_x509_crl *crl = (struct s2n_x509_crl*)(void*) mem.data;
+    crl->crl = NULL;
+
+    ZERO_TO_DISABLE_DEFER_CLEANUP(mem);
+    return crl;
+}
+
+int s2n_x509_crl_from_pem(char *pem, struct s2n_x509_crl **crl) {
```

Comment:
From an offline discussion: Avoid cstrings / use explicit lengths, and see if you can avoid allocating as much memory.

Reply:
I tried to perform less allocations here. I think the der stuffer and the crl blob still need to be allocated since the pem is being converted to a der in `s2n_stuffer_crl_from_pem`.